onnxruntime/core/providers/webgpu/nn/pool.cc

```diff
@@ -0,0 +1,254 @@
+POOLING_KERNEL_VERSIONED(AveragePool, kOnnxDomain, false, AveragePool, 10, 10)
+POOLING_KERNEL_VERSIONED(AveragePool, kMSInternalNHWCDomain, true, AveragePool, 10, 10)
+POOLING_KERNEL(AveragePool, kOnnxDomain, false, AveragePool, 11)
+POOLING_KERNEL(AveragePool, kMSInternalNHWCDomain, true, AveragePool, 11)
+POOLING_KERNEL(GlobalAveragePool, kOnnxDomain, false, AveragePool, 1)
+POOLING_KERNEL(GlobalAveragePool, kMSInternalNHWCDomain, true, AveragePool, 1)
+
+POOLING_KERNEL_VERSIONED(MaxPool, kOnnxDomain, false, MaxPool<1>, 1, 7)
+POOLING_KERNEL_VERSIONED(MaxPool, kMSInternalNHWCDomain, true, MaxPool<1>, 1, 7)
+POOLING_KERNEL_VERSIONED_WITH_INDICES(MaxPool, kOnnxDomain, false, MaxPool<8>, 8, 9)
+POOLING_KERNEL_VERSIONED_WITH_INDICES(MaxPool, kMSInternalNHWCDomain, true, MaxPool<8>, 8, 9)
+POOLING_KERNEL_VERSIONED_WITH_INDICES(MaxPool, kOnnxDomain, false, MaxPool<8>, 10, 10)
+POOLING_KERNEL_VERSIONED_WITH_INDICES(MaxPool, kMSInternalNHWCDomain, true, MaxPool<8>, 10, 10)
+POOLING_KERNEL_VERSIONED_WITH_INDICES(MaxPool, kOnnxDomain, false, MaxPool<8>, 11, 11)
+POOLING_KERNEL_VERSIONED_WITH_INDICES(MaxPool, kMSInternalNHWCDomain, true, MaxPool<8>, 11, 11)
+POOLING_KERNEL_WITH_INDICES(MaxPool, kOnnxDomain, false, MaxPool<8>, 12)
+POOLING_KERNEL_WITH_INDICES(MaxPool, kMSInternalNHWCDomain, true, MaxPool<8>, 12)
+POOLING_KERNEL(GlobalMaxPool, kOnnxDomain, false, MaxPool<1>, 1)
+POOLING_KERNEL(GlobalMaxPool, kMSInternalNHWCDomain, true, MaxPool<1>, 1)
+
+Status PoolProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  const auto& input = shader.AddInput("input", ShaderUsage::UseUniform);
+  const auto& output = shader.AddOutput("output", ShaderUsage::UseUniform);
+
+  // Declare and initialize the variables needed.
+  std::string var_decl_code;
+  // Process each element in the pooling window.
+  std::string sampling_code;
+  // Calculate the output value for each pooling window.
+  std::string downsampling_code;
+  if (is_max_pool_) {
+    std::string f16_min = "f16(-65504)";
+
+    std::stringstream f32_min_ss;
+    f32_min_ss << "f32(" << std::numeric_limits<float>::lowest() << ")";
+    std::string f32_min = f32_min_ss.str();
+
+    std::stringstream var_decl_ss;
+    var_decl_ss << "  var value = " << (is_float16_ ? f16_min : f32_min) << ";\n";
+    var_decl_code = var_decl_ss.str();
+
+    sampling_code = "      value = max(value, x_val);\n";
+  } else {
+    std::stringstream var_decl_ss;
+    var_decl_ss << "  var value = " << (is_float16_ ? "f16(0)" : "f32(0)") << ";\n";
+    if (!count_include_pad_) {
+      var_decl_ss << "  var count = u32(0);\n";
+    } else {
+      var_decl_ss << "  var count = uniforms.kernel_size;\n";
+    }
+    var_decl_code = var_decl_ss.str();
+
+    std::stringstream sampling_ss;
+    sampling_ss << "      value += x_val;\n";
+    if (!count_include_pad_) {
+      sampling_ss << "      count++;\n";
+    }
+    sampling_code = sampling_ss.str();
+
+    std::stringstream downsampling_ss;
```

Comment:
`std::stringstream` is heavy. please check string_macros.h to use macro `SS` which uses `onnxruntime::webgpu::OStringStream`(an alias of `absl::strings_internal::OStringStream`) with a pre-allocated size to minimize memory allocation.

Reply:
Done.